code/layout-analyzer.js

```diff
@@ -202,210 +341,123 @@ window.addEventListener('DOMContentLoaded', () => {
-          } else if (curr_finger === last_finger) {
-            // same finger
-            sfuDigrams[digram] = frequency;
-            sfuCount[curr_finger] += frequency;
-          } else if (curr_finger[0] === last_finger[0]) {
-            // same hand
-            // if (Math.abs(getKeyRow(key.id) - getKeyRow(acc)) >= 2) {
-            if (isScisor(key.id, acc, curr_finger, last_finger)) {
-              scisors[digram] = frequency;
-
-              if (requiresExtension(key.id) || requiresExtension(acc)) {
-                extendedRolls[digram] = frequency;
-              }
-            } else {
-              if (requiresExtension(key.id) || requiresExtension(acc)) {
-                extendedRolls[digram] = frequency;
-              } else if (curr_finger[1] < last_finger[1]) {
-                inwardDigrams[digram] = frequency;
-              } else {
-                outwardDigrams[digram] = frequency;
-              }
-            }
-          }
+    const getDigramType = (prevKeyCode, currKeyCode) => {
+      if (prevKeyCode === currKeyCode) return 'skb';
 
-          return key.id;
-        }, '');
-      });
+      const prevFinger = keyFinger[prevKeyCode];
+      const currFinger = keyFinger[currKeyCode];
 
-    document.querySelector('#imprecise-data').style.display = impreciseData
-      ? 'block'
-      : 'none';
+      if (currFinger === prevFinger) return 'sfb';
+      if (currFinger[0] !== prevFinger[0]) return 'handChange';
 
-    const sumUpFrequencies = dict => Object.values(dict).reduce(sum, 0);
+      if (isScisor(currKeyCode, prevKeyCode, currFinger, prevFinger))
+        return [prevKeyCode, currKeyCode].some(requiresExtension)
+          ? 'extendedScisor'
+          : 'scisor';
 
-    const mergeSfuSku = (sfu, sku, baseIndex) => {
-      const rv = [];
-      for (let i = baseIndex; i < baseIndex + 4; i++) {
-        rv.push({ "good": 0, "meh": sku[i], "bad": sfu[i] });
-      }
-      return rv;
+      if ([prevKeyCode, currKeyCode].some(requiresExtension)) return 'lsb';
+      return currFinger[1] < prevFinger[1] ? 'inwardRoll' : 'outwardRoll';
     };
 
-    const badDigrams = [
-      mergeSfuSku(Object.values(sfuCount), Object.values(skuCount), 0),
-      mergeSfuSku(Object.values(sfuCount), Object.values(skuCount), 4),
-    ];
+    const getTrigramType = (prevKeyCode, currKeyCode, nextKeyCode) => {
+      const prevFinger = keyFinger[prevKeyCode];
+      const currFinger = keyFinger[currKeyCode];
+      const nextFinger = keyFinger[nextKeyCode];
+
+      if (prevFinger == nextFinger) return prevKeyCode == nextKeyCode ? 'sks' : 'sfs';
+
+      const hands = prevFinger[0] + currFinger[0] + nextFinger[0];
+
+      if (!['lll', 'rrr'].includes(hands)) return 'other';
+
+      const fingers = prevFinger[1] + currFinger[1] + nextFinger[1];
+      if (fingers[0] == fingers[1] || fingers[1] == fingers[2]) return 'other';
+
+      // I don’t know if this is briliant or if I deserve a VIP ticket to Hell.
```

Comment:
I’m quite confident Hell is full of brilliant people. You’ll feel at home.

Reply:
Alright, I’ll see if I can rewrite that 😅

Reply:
Nah it’s fine as it is. Especially with this comment.
Marking as resolved.